File: .reference/monty/crates/monty-python/python/pydantic_monty/_binary.py
"""Locates the `monty` CLI binary that worker pools run as subprocesses.

The binary ships in the `pydantic-monty-runtime` wheel (a maturin `bin`-bindings
package, the same pattern `uv` and `ruff` use), which installs it into the
environment's scripts directory. Installing `pydantic-monty` pulls that in; the
`pydantic-monty-client` distribution alone does not, so the other resolution
steps below matter when only the bindings are installed.
"""

from __future__ import annotations

import os
import shutil
import sys
import sysconfig
from pathlib import Path

_BIN_NAME = 'monty.exe' if sys.platform == 'win32' else 'monty'


def find_monty_binary(explicit: str | Path | None = None) -> str:
    """Resolves the `monty` binary path used to spawn pool workers.

    Resolution order:

    1. the explicit `binary_path` argument, when given
    2. the `MONTY_BIN` environment variable
    3. the environment's scripts directory (where the `pydantic-monty-runtime`
       wheel installs the binary)
    4. a `monty` executable on `PATH`
    5. a cargo-built binary in the monty workspace, when running from an
       editable/development install of this package
    """
    if explicit is not None:
        return str(explicit)
    env = os.environ.get('MONTY_BIN')
    if env:
        return env
    for scripts_dir in _script_dirs():
        candidate = scripts_dir / _BIN_NAME
        if candidate.is_file():
            return str(candidate)
    found = shutil.which('monty')
    if found:
        return found
    dev = _development_binary()
    if dev is not None:
        return str(dev)
    raise FileNotFoundError(
        'could not locate the `monty` binary required to run sandboxed code; '
        'install it with `pip install pydantic-monty` (or `make dev-py` in the monty repo), '
        'pass binary_path=..., or set the MONTY_BIN environment variable'
    )


def _script_dirs() -> list[Path]:
    """Scripts directories of the current environment, most specific first."""
    dirs = [Path(sysconfig.get_path('scripts'))]
    # user-scheme installs (`pip install --user`) place scripts elsewhere
    try:
        dirs.append(Path(sysconfig.get_path('scripts', f'{os.name}_user')))
    except KeyError:
        pass
    return dirs


def _development_binary() -> Path | None:
    """A cargo-built `monty` next to an editable install inside the monty repo.

    With `maturin develop`, this file lives at
    `<repo>/crates/monty-python/python/pydantic_monty/_binary.py`, so a
    cargo-built binary is at `<repo>/target/{debug,release}/monty`. Prefer the
    most recently built one so `cargo build` and `cargo build --release`
    both behave as expected.
    """
    repo = Path(__file__).resolve().parents[4]
    if not (repo / 'Cargo.toml').is_file():
        return None
    candidates = [path for profile in ('debug', 'release') if (path := repo / 'target' / profile / _BIN_NAME).is_file()]
    return max(candidates, key=lambda path: path.stat().st_mtime, default=None)

File: .reference/monty/crates/monty-js/ts/binary.ts
// Locates the `monty` CLI binary that worker subprocesses run.
//
// Resolution order mirrors pydantic_monty's `_binary.py`:
// 1. an explicit `binaryPath` option,
// 2. the `MONTY_BIN` environment variable,
// 3. the platform-specific npm package (`@pydantic/monty-<platform>`,
//    installed automatically via optionalDependencies),
// 4. `monty` on PATH,
// 5. a cargo workspace `target/{debug,release}` build (development fallback).

import { accessSync, constants, existsSync, statSync } from 'node:fs'
import { createRequire } from 'node:module'
import { delimiter, dirname, join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const EXE = process.platform === 'win32' ? 'monty.exe' : 'monty'

/**
 * The napi-style platform triple used to name binary packages, or `null` on
 * platforms we do not ship binaries for.
 */
export function platformTriple(): string | null {
  const { platform, arch } = process
  if (platform === 'darwin' && (arch === 'x64' || arch === 'arm64')) {
    return `darwin-${arch}`
  }
  if (platform === 'linux' && (arch === 'x64' || arch === 'arm64')) {
    return `linux-${arch}-gnu`
  }
  if (platform === 'win32' && arch === 'x64') {
    return 'win32-x64-msvc'
  }
  return null
}

/**
 * Resolves the `monty` binary path, throwing a descriptive error naming
 * every location tried when nothing is found.
 */
export function findMontyBinary(explicit?: string): string {
  if (explicit !== undefined) {
    if (!existsSync(explicit)) {
      throw new Error(`monty binary not found at binaryPath: ${explicit}`)
    }
    return explicit
  }

  const tried: string[] = []

  const envBin = process.env.MONTY_BIN
  if (envBin) {
    if (existsSync(envBin)) {
      return envBin
    }
    tried.push(`MONTY_BIN=${envBin}`)
  }

  const fromPackage = platformPackageBinary()
  if (fromPackage !== null) {
    return fromPackage
  }
  tried.push('platform package @pydantic/monty-<platform>')

  const fromPath = searchPath()
  if (fromPath !== null) {
    return fromPath
  }
  tried.push('PATH')

  const fromWorkspace = workspaceBinary()
  if (fromWorkspace !== null) {
    return fromWorkspace
  }
  tried.push('cargo workspace target/')

  throw new Error(
    `could not locate the monty binary (tried: ${tried.join(', ')}). ` +
      'Install the platform package, set MONTY_BIN, or pass binaryPath.',
  )
}

/**
 * The binary shipped by the platform-specific npm package, if installed.
 *
 * Resolution failures fall through to the next strategy rather than erroring:
 * the same package names previously shipped napi `.node` bindings, so a stale
 * install can resolve while holding no `monty` executable.
 */
function platformPackageBinary(): string | null {
  const triple = platformTriple()
  if (triple === null) {
    return null
  }
  const require = createRequire(import.meta.url)
  try {
    return require.resolve(`@pydantic/monty-${triple}/${EXE}`)
  } catch (err) {
    // Only not-found-style failures fall through (package not installed, or
    // installed without the binary). Anything else — permission errors,
    // malformed package.json — is a real problem the user must see, not a
    // cue to silently pick a different binary.
    const code = (err as NodeJS.ErrnoException).code
    if (code === 'MODULE_NOT_FOUND' || code === 'ERR_MODULE_NOT_FOUND' || code === 'ERR_PACKAGE_PATH_NOT_EXPORTED') {
      return null
    }
    throw err
  }
}

/** Scans PATH directories for an executable `monty`. */
function searchPath(): string | null {
  for (const dir of (process.env.PATH ?? '').split(delimiter)) {
    if (dir === '') {
      continue
    }
    const candidate = join(dir, EXE)
    if (isExecutableFile(candidate)) {
      return candidate
    }
  }
  return null
}

/** Whether `path` is a regular file the current process may execute. */
function isExecutableFile(path: string): boolean {
  try {
    if (!statSync(path).isFile()) {
      return false
    }
    // X_OK is meaningless on Windows, where any readable file is "executable"
    accessSync(path, process.platform === 'win32' ? constants.R_OK : constants.X_OK)
    return true
  } catch {
    return false
  }
}

/**
 * Development fallback: walk up from this file looking for a cargo workspace
 * containing a built `monty` binary (debug preferred — it matches the code
 * being developed; release as fallback).
 */
function workspaceBinary(): string | null {
  let dir = dirname(fileURLToPath(import.meta.url))
  for (let i = 0; i < 6; i++) {
    if (existsSync(join(dir, 'Cargo.toml'))) {
      for (const profile of ['debug', 'release']) {
        const candidate = join(dir, 'target', profile, EXE)
        if (existsSync(candidate)) {
          return candidate
        }
      }
    }
    const parent = resolve(dir, '..')
    if (parent === dir) {
      break
    }
    dir = parent
  }
  return null
}
